Feature/group api (#91)
upb-uc4/api: hlf/chaincode/group.md

# Hyperledger Group Api

In the following, the chaincode api for the "Group"-Chaincode is described.
This contains all Transactions and Models provided for this domain.
The Errors returned are defined [here](errors.md#Errors).

## ContractName "UC4.Group"

## Transactions

### Add User To Group
- ID = addUserToGroup
- Send
    - enrollmentID :: String

>     - groupID :: String

Reviewer: ```suggestion
    - enrollmentId :: String
    - groupId :: String
```

- Receive
    - group :: [Group](#Group)

>         - Description: ledgerState after transaction.

Reviewer: ```suggestion
    - ""
      -  Description: Done; returns empty String on success.
```

> - Receive
>     - userList :: List\<String\>
>         - Description: ledgerState after transaction.

Author: ```suggestion
- Receive
    - group :: [Group](#Group)
        - Description: ledgerState after transaction.
```

    - [DetailedError](errors.md#DetailedError) 
      ```json
      {
        "type": "HLUnprocessableEntity",
        "title": "The following parameters do not conform to the specified format",
        "invalidParams": [
          {
            "name": "string",
            "reason": "string"
          }
        ]
      }
      ```
       - Description: This error is returned, 
         - if the given parameters could not be parsed. If some attributes within the json are not well formatted, they are listed in invalidParams.  
            For detailed informations see [parameter checks](#parameterChecks).
         - if the given parameters produce some [semantic error](#semanticChecks).
    
    - [GenericError](errors.md#GenericError) 
      ```json
      {
        "type": "HLInsufficientApprovals",
        "title": "The approvals present on the ledger do not suffice to execute this transaction"
      }
      ```
      - Description: This error is returned, if the required approvals are not present.

### Remove From Group
- ID = removeUserFromGroup
- Send
    - enrollmentID :: String

>     - groupID :: String

Reviewer: ```suggestion
    - enrollmentId :: String
    - groupId :: String
```

- Receive
    - userList :: List\<String\>

>             - Description: ledgerState after transaction.

Reviewer: ```suggestion
    - ""
      -  Description: Done; returns empty String on success.
```

      }
      ```
      - Description: This error is returned, if there is no match for the given enrollmentId and groupId.

    - [DetailedError](errors.md#DetailedError) 
      ```json
      {
        "type": "HLUnprocessableEntity",
        "title": "The following parameters do not conform to the specified format",
        "invalidParams": [
          {
            "name": "string",
            "reason": "string"
          }
        ]
      }
      ```
      - Description: This error is returned, if the given parameters are not well formatted, they are listed in invalidParams.  
      For detailed informations see [parameter checks](#parameterChecks).
      (enrollmentId, groupId)

    - [GenericError](errors.md#GenericError) 
      ```json
      {
        "type": "HLInsufficientApprovals",
        "title": "The approvals present on the ledger do not suffice to execute this transaction"
      }
      ```
      - Description: This error is returned, if the required approvals are not present.

### Remove User From All Groups
- ID = removeUserFromAllGroups
- Send
    - enrollmentId :: String
- Receive
    - ""

>       -  Description: Done, returns the submitted data.

Reviewer: ```suggestion
      -  Description: Done; returns empty String on success.
```

      }
      ```
      - Description: This error is returned, if the state of data on the ledger is not consistent with the curent model. This error should only occurr if the model changes while the old ledger state remains without modification.

    - [GenericError](errors.md#GenericError) 
      ```json
      {
        "type": "HLNotFound",
        "title": "There is no group for the given groupId."
      }
      ```
      - Description: This error is returned, if the groupId does not exist on the ledger.

    - [DetailedError](errors.md#DetailedError) 
      ```json
      {
        "type": "HLUnprocessableEntity",
        "title": "The following parameters do not conform to the specified format",
        "invalidParams": [
          {
            "name": "string",
            "reason": "string"
          }
        ]
      }
      ```
      - Description: This error is returned, if the given parameters are not well formatted, they are listed in invalidParams.  
      For detailed informations see [parameter checks](#parameterChecks).

### Get Groups For User
- ID = getGroupsForUser
- Send
    - enrollmentId :: String

- Receive

>     - groupList :: List\<String\>

Reviewer: ```suggestion
    - groupIdList :: List\<String\>
```

>           - Description: Exhaustive List of all group names the user is part of.

Reviewer: ```suggestion
          - Description: Exhaustive List of all groupIds the user is part of.
```

    - [GenericError](errors.md#GenericError) 
      ```json
      {
        "type": "HLUnprocessableLedgerState",
        "title": "The state on the ledger does not conform to the specified format"
      }
      ```
      - Description: This error is returned, if the state of data on the ledger is not consistent with the curent model. This error should only occurr if the model changes while the old ledger state remains without modification.

    - [DetailedError](errors.md#DetailedError) 
      ```json
      {
        "type": "HLUnprocessableEntity",
        "title": "The following parameters do not conform to the specified format",
        "invalidParams": [
          {
            "name": "string",
            "reason": "string"
          }
        ]
      }
      ```
      - Description: This error is returned, if the given parameters are not well formatted, they are listed in invalidParams. It is also returned, if the user is not registered.  
      For detailed informations see [parameter checks](#parameterChecks) and [semanticChecks](#semanticChecks).

## <a id="Models" />Models

### <a id="Group" />Group
```json
{
  "groupId": "0123456ExampleGroup",

>   "userList": ["0123456", "123"]

Reviewer: ```suggestion
  "userList": ["EnrollmentIdExample123", "EnrollmentIdExample124"]
```
